pkg/retrieval/retrieval.go

```diff
@@ -112,134 +112,88 @@ func (s *Service) RetrieveChunk(ctx context.Context, addr swarm.Address, origin
 
 		sp := newSkipPeers()
 
-		ticker := time.NewTicker(retrieveRetryIntervalDuration)
-		defer ticker.Stop()
-
 		var (
-			peerAttempt  int
-			peersResults int
-			resultC      = make(chan retrievalResult, maxSelects)
+			allowedRetries   = maxRetries
+			allowedRetrieves = 1
 		)
 
-		requestAttempt := 0
+		if origin {
+			allowedRetrieves = maxRetrieved
+		}
+
+		timer := time.NewTimer(0)
+		defer timer.Stop()
 
-		lastTime := time.Now().Unix()
+		resultChan := make(chan retrievalResult)
+		doneChan := make(chan struct{})
+		defer close(doneChan)
 
-		for requestAttempt < maxRequestRounds {
+		for {
 
-			if peerAttempt < maxSelects {
+			select {
+			case <-ctx.Done():
+				return nil, storage.ErrNotFound
+			case <-timer.C:
+
+				allowedRetries--
+				allowedRetrieves--
+
+				s.metrics.PeerRequestCounter.Inc()
 
 				// create a new context without cancelation but
 				// set the tracing span to the new context from the context of the first caller
-				ctx := tracing.WithContext(context.Background(), tracing.FromContext(topCtx))
+				ctx := tracing.WithContext(context.Background(), tracing.FromContext(ctx))
 
 				// get the tracing span
 				span, _, ctx := s.tracer.StartSpanFromContext(ctx, "retrieve-chunk", s.logger, opentracing.Tag{Key: "address", Value: addr.String()})
 				defer span.Finish()
 
-				peerAttempt++
-				s.metrics.PeerRequestCounter.Inc()
 				go func() {
-
 					// cancel the goroutine just with the timeout
-					ctx, cancel := context.WithTimeout(ctx, retrieveChunkTimeout)
+					ctx, cancel := context.WithTimeout(ctx, defaultTTL)
 					defer cancel()
-
-					chunk, peer, requested, err := s.retrieveChunk(ctx, addr, sp, origin)
-					select {
-					case resultC <- retrievalResult{
-						chunk:     chunk,
-						peer:      peer,
-						err:       err,
-						retrieved: requested,
-					}:
-					case <-ctx.Done():
-					}
-
+					s.retrieveChunk(ctx, resultChan, doneChan, addr, sp, origin)
 				}()
-			} else {
-				select {
-				case resultC <- retrievalResult{}:
-				case <-ctx.Done():
+
+				if allowedRetries <= 0 || allowedRetrieves <= 0 {
+					continue
 				}
-			}
 
-			select {
-			case <-ticker.C:
-				// break
-			case res := <-resultC:
+				timer.Reset(p90TTL)
+
+			case res := <-resultChan:
+
 				if errors.Is(res.err, topology.ErrNotFound) {
 					if sp.Saturated() {
 						// if no peer is available, and none skipped temporarily
 						s.logger.Tracef("retrieval: failed to get chunk %s", addr)
 						return nil, storage.ErrNotFound
-					} else {
-						// skip to next request round if any peers are only skipped temporarily
-						peerAttempt = maxSelects
 					}
 				}
+
 				if res.retrieved {
-					if res.err != nil {
-						if !res.peer.IsZero() {
-							s.logger.Debugf("retrieval: failed to get chunk %s from peer %s: %v", addr, res.peer, res.err)
-						}
-						peersResults++
-					} else {
+					if res.err == nil {
```

Comment:
maybe just checking whether the error is nil would be enough? can't `retrieved` be assumed from that?